### PR TITLE
PRWM Loader

### PR DESCRIPTION
**Pull request by kchapelier**
This PR implements a loader for Three.js, includes a few PRWM models (vive controller, a smooth suzanne and a facetted nefertiti from http://nefertitihack.alloversky.com/ which is public domain) and an simple example. A similar example with more models is available [here](http://www.kchapelier.com/prwm/examples/three-prwm-loader.html) and a little benchmark comparing some loaders is available [here](http://www.kchapelier.com/prwm/examples/three-prwm-loader-benchmark.html).

There's also a `three-buffergeometry-to-prwm` npm module which takes an instance of Three.BufferGeometry as argument and return a PRWM file as an arraybuffer ([online example](http://www.kchapelier.com/prwm/examples/three-buffergeometry-to-prwm.html)). I intend to work on a proper Three.js exporter (like OBJExporter.js) at some point.

----

PRWM is a binary file format for nD geometries specifically designed for JavaScript and WebGL with a strong focus on fast parsing (from 1ms to less than 0.1ms in Chrome 59 on a MBP Late 2013). It achieves its speed by using typed array as windowed views over the arraybuffer retrieved in ajax, this means that the loader is not actually reading the array buffers for the attributes and the indices, just *pointing* at them. Additionally the size of the file being parsed has no influence on the parsing time.

The format supports indexed and non-indexed geometries, as well as custom attributes making it is possible to store custom geomeries such as the one for Jaume Sanchez's Three.MeshLine. Size-wise a PRWM file is usually between the equivalent ascii OBJ and binary FBX files. Models with smooth surfaces can usually shave off a lot of bytes when they're saved as indexed geometries, I was able to convert a 1.6MB OBJ file to a 432KB PRWM file for a project I'm currently working on.

I wrote slightly more about all this on https://github.com/kchapelier/PRWM if you're interested in more gritty details and the file format specs.

### CODE REVIEW

**Comment by mrdoob** [comment]
> The parsing of PRWM file is especially fast when the endianness of the file is the same as the endianness of the client platform. 

Interesting. I wonder if there is a better way to handle this. Say, naming the files like this:

```
cerberus.be.prwm
cerberus.le.prwm
```

Then, using the loader like `load('cerberus.prwm')` and letting the loader look for `be.pwrn` or `le.prwm` depending of the user's endianness.

> There's also a `three-buffergeometry-to-prwm` npm module which takes an instance of Three.BufferGeometry as argument and return a PRWM file as an arraybuffer ([online example](http://www.kchapelier.com/prwm/examples/three-buffergeometry-to-prwm.html)).

I guess this only outputs a prwm in the user's endianness?

**Comment by kchapelier** [comment]
> Interesting. I wonder if there is a better way to handle this. Say, naming the files like this:
> ```js 
> cerberus.be.prwm
> cerberus.le.prwm
> ```
> Then, using the loader like load('cerberus.prwm') and letting the loader look for be.pwrm or le.prwm depending of the user's endianness.

I thought about it but I'm not sure this is the way to go.. While this would reduce the number of end points on the API, I think this would make it less "straightforward" to use. If possible I'd rather avoid forcing the users to set filenames following a specific format.

Currently the loader can load files of any endianness, it's just much faster if it matches your plaftorm's. But in my tests parsing a PRWM file of the "wrong" endianness is still slightly faster than parsing an equivalent OBJ model so it should still be allowed. To allow that in this example we'd need to first check if the `cerberus.prwm` file exists, then if we get a 404 try to access `cerberus.le.prwm`, or check for `cerberus.le.prwm` and if it does not exist try to access `cerberus.be.prwm`.

Currently, there's a loadBest method with this signature : 

```js
loader.loadBest( urlLittleEndian, urlBigEndian, onLoad, onProgress, onError )
```

which use one or the other depending on the client platform endianness , and for users who'd want to handle that themselves : 

```js
THREE.PRWMLoader.isBigEndianPlatform()
```

One possibility to simplify the API would be to *optionally* accept an array (with the littleEndian url and the bigEndian url) instead of a string as the first argument of `prwmLoader.load`. Not sure I like methods with "flexible" signatures though...

> > There's also a three-buffergeometry-to-prwm npm module which takes an instance of Three.BufferGeometry as argument and return a PRWM file as an arraybuffer ([online example](http://www.kchapelier.com/prwm/examples/three-buffergeometry-to-prwm.html)).
> 
> I guess this only outputs a prwm in the user's endianness?

The example is hardcoded to generate little endian files, but the choice of the endianness is exposed by the actual module API.

```js
threeBufferGeometryToPrwm(bufferGeometry, bigEndian);
```

**Comment by mrdoob** [comment]
> To allow that in this example we'd need to first check if the `cerberus.prwm` file exists, then if we get a 404 try to access `cerberus.le.prwm`, or check for `cerberus.le.prwm` and if it does not exist try to access `cerberus.be.prwm`.

Actually, what I'm proposing is that, if the user tries to load `cerberus.prwm`, the loader will check `isBigEndianPlatform()` and, depending of the boolean, load `cerberus.be.prwm` or `cerberus.le.prwm`.

So, they'll always have to provide 2 files per model, but I think the parse speed is well worth it.

**Comment by mrdoob** [comment]
Maybe this would be clearer:

```js
loader.load( 'cerberus.*.prwm', function () {} );
```

That way the loader just needs to replace the `*`.

**Comment by kchapelier** [comment]
Yes good idea, this is is more straightforward and allow to bypass the filename change by not using `*`. I'll implement those changes tomorrow.

**Comment by kchapelier** [comment]
The change discussed above has been implemented, the `loadBest` method removed and the example adapted. `THREE.PRWMLoader.isBigEndianPlatform()` is still exposed in case someone want to do a wrapper for the loader for some reason.

File: examples/js/loaders/PRWMLoader.js [hunk 1]
```diff
@@ -0,0 +1,293 @@
...
+
+			console.timeEnd( 'PRWMLoader' );
+
+			return bufferGeometry;
+
+		}
+
+	};
+
+	THREE.PRWMLoader.isBigEndianPlatform = function () {
```
Comment (mrdoob) [inline examples/js/loaders/PRWMLoader.js]: This is probably not longer needed?

Comment (mrdoob) [inline examples/js/loaders/PRWMLoader.js]: This is probably no longer needed?

Comment (mrdoob) [inline examples/js/loaders/PRWMLoader.js]: Never mind. Just saw your comment.

**Comment by mrdoob** [comment]
Thanks!

**Comment by mrdoob** [comment]
While talking about this PR with @spite, he raised an valid question. Are there any big-endian platforms that run WebGL?

**Comment by kchapelier** [comment]
> While talking about this PR with @spite, he raised an valid question: Are there any big-endian platforms that run WebGL?

That's hard to say to be honest. My researches lead me to think that exotic big-endian android devices and custom big-endian *nix box might exist in the wild due to the bi-endianness of arm/aarch64 CPUs (even though they're *typically* set to little-endian), but it's hard to find conclusive evidence of this.

**Comment by mrdoob** [comment]
I'll ask around tomorrow, but it's tempting to simplify it even more and simplify it to just little-endian...

**Pull request merged**